metricbeat/module/prometheus/query/data.go

```diff
@@ -0,0 +1,73 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package query
+
+import (
+	"encoding/json"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/metricbeat/mb"
+)
+
+// PromResponseBody for Prometheus Query API Request
+type PromResponseBody struct {
+	Status string         `json:"status"`
+	Data   prometheusData `json:"data"`
+}
+type prometheusData struct {
+	ResultType string   `json:"resultType"`
+	Results    []result `json:"result"`
+}
+type result struct {
+	Metric          interface{}            `json:"metric"`
+	Vectors         []interface{}          `json:"value,omitempty"`
+	ReconciledValue map[string]interface{} `json:"reconciledValue"`
+}
+
+func (m *MetricSet) parseResponse(body []byte, pathConfig PathConfig) mb.Event {
+	var event common.MapStr
+	var res PromResponseBody
+	if err := json.Unmarshal(body, &res); err != nil {
+		m.Logger().Error("Failed to parsing api response ", err)
+	}
+
+	// Check if there is vector array.
+	// Vector [ <unix_timestamp>, "<query_result>" ] is not acceptable for Elasticsearch.
+	// Because there are two types in one array.
+	// So change Vector to Object { unixtimestamp: "<unix_timestamp", value: "query_result" }
+	if res.Data.ResultType == "vector" {
```

Comment:
It would be nice to come up with a good conversion for all result types. I think we can strip down much of the info we get in the response to store only timestamp + results. What do you think about doing something like this?:

- For `scalar`:
    - Store timestamp under event tiemstamp (`mb.Event.Timestamp`) (this will be reported as `@timestamp` in the final event)
    - Store value under `prometheus.query.<query_name>` (ie: `prometheus.query.http_request` in your example)
- For `string`:
    - Store timestamp under event tiemstamp (`mb.Event.Timestamp`)
    - Store value under `prometheus.query.<query_name>`
- For `vector`:
    - Create an event for each metric
    - Store timestamp under event timestamp (`mb.Event.Timestamp`)
    - Store value under `prometheus.query.<query_name>`
    - Store labels under `prometheus.labels`
- For `matrix` (range vector):
    - Create an event for each metric **and** (timestamp, vlue) pairs
    - Store timestamp under event timestamp (`mb.Event.Timestamp`)
    - Store value under `prometheus.query.<query_name>`
    - Store labels under `prometheus.labels`

Reply:
Looks good! I'll try to apply this.

Reply:
I'm applying the suggested changes on code. Please review commit [#c511e95faf](https://github.com/elastic/beats/pull/15177/commits/c511e95faf834fdb75dbff040c77d4ad76475db4)